node/node-core/src/eth/receipt.rs

```diff
@@ -0,0 +1,161 @@
+use crate::eth::utils::enveloped;
+use ethers_core::{
+    types::{Address, Bloom, Bytes, H256, U256},
+    utils::{
+        rlp,
+        rlp::{Decodable, DecoderError, Encodable, Rlp, RlpStream},
+    },
+};
+use foundry_evm::revm;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct Log {
+    pub address: Address,
+    pub topics: Vec<H256>,
+    pub data: Bytes,
+}
+
+impl From<revm::Log> for Log {
+    fn from(log: revm::Log) -> Self {
+        let revm::Log { address, topics, data } = log;
+        Log { address, topics, data: data.into() }
+    }
+}
+
+impl From<Log> for revm::Log {
+    fn from(log: Log) -> Self {
+        let Log { address, topics, data } = log;
+        revm::Log { address, topics, data: data.0 }
+    }
+}
+
+impl rlp::Encodable for Log {
+    fn rlp_append(&self, stream: &mut rlp::RlpStream) {
+        stream.begin_list(3);
+        stream.append(&self.address);
+        stream.append_list(&self.topics);
+        stream.append(&self.data.as_ref());
+    }
+}
+
+impl rlp::Decodable for Log {
+    fn decode(rlp: &rlp::Rlp) -> Result<Self, rlp::DecoderError> {
+        let result = Log {
+            address: rlp.val_at(0)?,
+            topics: rlp.list_at(1)?,
+            data: rlp.val_at::<Vec<u8>>(2)?.into(),
+        };
+        Ok(result)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct EIP658Receipt {
+    pub status_code: u8,
+    pub gas_used: U256,
+    pub logs_bloom: Bloom,
+    pub logs: Vec<Log>,
+}
+
+impl rlp::Encodable for EIP658Receipt {
+    fn rlp_append(&self, stream: &mut rlp::RlpStream) {
+        stream.begin_list(4);
+        stream.append(&self.status_code);
+        stream.append(&self.gas_used);
+        stream.append(&self.logs_bloom);
+        stream.append_list(&self.logs);
+    }
+}
+
+impl rlp::Decodable for EIP658Receipt {
+    fn decode(rlp: &rlp::Rlp) -> Result<Self, rlp::DecoderError> {
+        let result = EIP658Receipt {
+            status_code: rlp.val_at(0)?,
+            gas_used: rlp.val_at(1)?,
+            logs_bloom: rlp.val_at(2)?,
+            logs: rlp.list_at(3)?,
+        };
+        Ok(result)
+    }
+}
+
+// same underlying data structure
+pub type EIP2930Receipt = EIP658Receipt;
+pub type EIP1559Receipt = EIP658Receipt;
```

Comment:
sorta wonder if we can get rid of these boilerplates? only difference seems to be the type right? should we just add a type field to EIP658Receipt and call it a day?

Reply:
I kinda like the enum approach which is in line with the other `Typed*` types and these alias are just to highlight this.